crates/control/src/role_assignment.rs

```diff
@@ -662,114 +687,66 @@ fn pick_role_with_penalties(
-    } else if unassigned_robots > 0
-        && role_assignment[PlayerNumber::Three].is_none()
-        && penalties[PlayerNumber::Three].is_none()
-    {
-        role_assignment[PlayerNumber::Three] = Some(Role::DefenderRight);
-        unassigned_robots -= 1;
-    } else if unassigned_robots > 0
-        && role_assignment[PlayerNumber::Four].is_none()
-        && penalties[PlayerNumber::Four].is_none()
-    {
-        role_assignment[PlayerNumber::Four] = Some(Role::DefenderRight);
-        unassigned_robots -= 1;
-    } else if unassigned_robots > 0
-        && role_assignment[PlayerNumber::Five].is_none()
-        && penalties[PlayerNumber::Five].is_none()
-    {
-        role_assignment[PlayerNumber::Five] = Some(Role::DefenderRight);
-        unassigned_robots -= 1;
-    }
+    // assign optional roles
+    let mut optional_role_index = 0;
 
-    if unassigned_robots > 0
-        && role_assignment[PlayerNumber::Two].is_none()
-        && penalties[PlayerNumber::Two].is_none()
-    {
-        role_assignment[PlayerNumber::Two] = Some(Role::StrikerSupporter);
-        unassigned_robots -= 1;
-    } else if unassigned_robots > 0
-        && role_assignment[PlayerNumber::Three].is_none()
-        && penalties[PlayerNumber::Three].is_none()
-    {
-        role_assignment[PlayerNumber::Three] = Some(Role::StrikerSupporter);
-        unassigned_robots -= 1;
-    } else if unassigned_robots > 0
-        && role_assignment[PlayerNumber::Four].is_none()
-        && penalties[PlayerNumber::Four].is_none()
-    {
-        role_assignment[PlayerNumber::Four] = Some(Role::StrikerSupporter);
-        unassigned_robots -= 1;
-    } else if unassigned_robots > 0
-        && role_assignment[PlayerNumber::Five].is_none()
-        && penalties[PlayerNumber::Five].is_none()
-    {
-        role_assignment[PlayerNumber::Five] = Some(Role::StrikerSupporter);
-        unassigned_robots -= 1;
+    if unassigned_robots > optional_roles.len() {
+        panic!("role_assignment not possible: not enough optional_roles available")
     }
 
-    if unassigned_robots > 0
-        && role_assignment[PlayerNumber::Two].is_none()
-        && penalties[PlayerNumber::Two].is_none()
-    {
-        role_assignment[PlayerNumber::Two] = Some(Role::DefenderLeft);
-    } else if unassigned_robots > 0
-        && role_assignment[PlayerNumber::Three].is_none()
-        && penalties[PlayerNumber::Three].is_none()
-    {
-        role_assignment[PlayerNumber::Three] = Some(Role::DefenderLeft);
-    } else if unassigned_robots > 0
-        && role_assignment[PlayerNumber::Four].is_none()
-        && penalties[PlayerNumber::Four].is_none()
-    {
-        role_assignment[PlayerNumber::Four] = Some(Role::DefenderLeft);
-    } else if unassigned_robots > 0
-        && role_assignment[PlayerNumber::Five].is_none()
-        && penalties[PlayerNumber::Five].is_none()
-    {
-        role_assignment[PlayerNumber::Five] = Some(Role::DefenderLeft);
+    while unassigned_robots > 0 {
```

Comment:
This should be a for loop over an iterator of the optional roles. I think about something like
`optional_roles.iter().take(unassigned_robots)`. Feel free to propose a solution using iterators or discuss it here.

Reply:
Using indices is often a characteristic in the code that possibly indicates a deeper problem

Reply:
done